src/common/hist_util.cc

```diff
@@ -204,75 +221,161 @@ void BuildHistKernel(const std::vector<GradientPair> &gpair,
+                            GHistRow hist, const GHistBuildingManager& hbm) {
+  const size_t size = row_indices.Size();
+  const size_t *rid = row_indices.begin;
+  auto const *pgh = reinterpret_cast<const float *>(gpair.data());
+  const BinIdxType *gradient_index = gmat.index.data<BinIdxType>();
+
+  auto const &row_ptr = gmat.row_ptr.data();
+  auto base_rowid = gmat.base_rowid;
+  const uint32_t *offsets = gmat.index.Offset();
+  auto get_row_ptr = [&](size_t ridx) {
+    return first_page ? row_ptr[ridx] : row_ptr[ridx - base_rowid];
+  };
+  auto get_rid = [&](size_t ridx) {
+    return first_page ? ridx : (ridx - base_rowid);
+  };
+
+  const size_t n_features = gmat.cut.Ptrs().size() - 1;
+  const size_t n_columns = GHistBuildingManager::kColumnSampling ? hbm.fids.size() : n_features;
+  auto hist_data = reinterpret_cast<double *>(hist.data());
+  const uint32_t two{2};  // Each element from 'gpair' and 'hist' contains
+                          // 2 FP values: gradient and hessian.
+                          // So we need to multiply each row-index/bin-index by 2
+                          // to work with gradient pairs as a singe row FP array
+  for (size_t cid = 0; cid < n_columns; ++cid) {
+    const size_t local_cid = GHistBuildingManager::kColumnSampling ? hbm.fids[cid] : cid;
+    for (size_t i = 0; i < size; ++i) {
+      const size_t row_id = rid[i];
+      const size_t icol_start =
+          any_missing ? get_row_ptr(row_id) : get_rid(row_id) * n_features;
+
+      const BinIdxType *gr_index_local = gradient_index + icol_start;
+      const uint32_t idx_bin = two * (static_cast<uint32_t>(gr_index_local[local_cid]) +
+                                      (any_missing ? 0 : offsets[local_cid]));
+      auto hist_local = hist_data + idx_bin;
+
+      const size_t idx_gh = two * row_id;
+      // The trick with pgh_t buffer helps the compiler to generate faster binary.
+      const float pgh_t[] = {pgh[idx_gh], pgh[idx_gh + 1]};
+      *(hist_local)     += pgh_t[0];
+      *(hist_local + 1) += pgh_t[1];
+    }
+  }
+}
+
+template <bool do_prefetch, typename BinIdxType, bool first_page,
+          bool any_missing, class GHistBuildingManager>
+void BuildHistKernel(const std::vector<GradientPair> &gpair,
+                     const RowSetCollection::Elem row_indices, const GHistIndexMatrix &gmat,
+                     GHistRow hist, const GHistBuildingManager& hbm) {
+  if (GHistBuildingManager::kReadByColumn) {
+    ColsWiseBuildHistKernel<BinIdxType, first_page, any_missing>
+                           (gpair, row_indices, gmat, hist, hbm);
+  } else {
+    RowsWiseBuildHistKernel<do_prefetch, BinIdxType, first_page, any_missing>
+                           (gpair, row_indices, gmat, hist);
+  }
+}
+
+template <bool do_prefetch, bool first_page, bool any_missing, class GHistBuildingManager>
+void BuildHistDispatch(const std::vector<GradientPair> &gpair,
```

Comment:
Can we use the `DispatchBinType` instead?

Reply:
done